Code improvements to Swift 2 branch

NachoSoto opened the pull request:
I made a bunch of comments in #96, and here's my proposed solution to them, plus a few other improvements I've found along the way. Let me know what you think.
All tests are still passing.

### Changes:
- `SSLCert` is now `private`.
- `WebSockets`: removed unnecessary `weak`s and simplified code.
- `SwiftRegex` does not need to be exposed.
- `SwiftRegex`: failing with `fatalError`.
- `SwiftRegex`: changed static method to `static` to avoid unnecessary dynamic dispatching.
- `SocketPacket`: using `Logger` instead of `print`.
- `SocketIOClient`: simplified conditional.
- `SocketIOClient`: remove unnecessary `optional`.
- `SocketIOClient`: removed unnecessary `weak self`.
- `SocketIOClient`: added assertion to check `timeoutAfter` is not negative.
- `SocketIOClient`: making a few fields `private`.
- Simplified `SocketEngine.swift`.

NachoSoto commented on SocketIOClientSwift/SocketEngine.swift:
```diff
@@ -350,9 +350,11 @@ public final class SocketEngine: NSObject, WebSocketDelegate {
-        for msg in postWait {
-            ws?.writeString(msg)
-        }
+		if let ws = ws {
```
This avoids looping through `postWait` if `ws` is `nil`.

lightsprint09 commented on SocketIOClientSwift/SocketEngine.swift:
```diff
@@ -629,26 +629,22 @@ public final class SocketEngine: NSObject, WebSocketDelegate {
-        guard pingInterval != nil else {
-            return
-        }
+        if let pingInterval = pingInterval {
```
I would use` guard let  pingInterval = pingInterval else { return }`

nuclearace replied:
I actually prefer using if let

lightsprint09 replied:
but you win one indentation. It makes sense in methods with long single lines of code

NachoSoto replied:
I prefer to avoid early returns that can be problematic if you don't notice them in the future. But I understand that's an stylistic choice (unlike some other things in this PR), so I'll chose whatever you guys prefer.

hankbao commented:
The `weak` remained in `WebSocket.connect()` method actually causes a nasty race condition problem. 

The WebSocket instance within SocketEngine is released in `createWebsocket` but may still be accessed in a global concurrent queue in `WebSocket.connect()`.

`let weakSelf = self` will try to unwrap and retain a deallocated object and crash the app.

nuclearace commented on SocketIOClientSwift/SocketEngine.swift:
```diff
@@ -350,8 +350,10 @@ public final class SocketEngine: NSObject, WebSocketDelegate {
     private func flushWaitingForPostToWebSocket() {
-        for msg in postWait {
-            ws?.writeString(msg)
+        if let ws = ws {
```
There's no need to break this out. `ws?.writeString(msg)` won't crash if ws is nil

NachoSoto replied:
This is a small optimization: there's no need to iterate through `postWait` if `ws` is `nil` anyway.

nuclearace replied:
Is the optimization really worth it? I'd prefer to keep as is, unless you've determined this is a bottleneck

nuclearace commented on SocketIOClientSwift/SocketEngine.swift:
```diff
@@ -525,8 +527,8 @@ public final class SocketEngine: NSObject, WebSocketDelegate {
 
                 if msg.characters.count != 0 {
                     // Be sure to capture the value of the msg
-                    dispatch_async(handleQueue) {[weak self, msg] in
```
Is there a reason you're removing the weak self? Generally when you dispatch async you don't want to create a strong reference

nuclearace commented on SocketIOClientSwift/SocketEngine.swift:
```diff
@@ -672,16 +668,16 @@ public final class SocketEngine: NSObject, WebSocketDelegate {
     Write a message, independent of transport.
     */
     public func write(msg: String, withType type: PacketType, withData data: [NSData]?) {
-        dispatch_async(emitQueue) {[weak self] in
```
I would keep this weak, there's no need to keep a strong reference.

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -148,28 +153,30 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
     public func connect(timeoutAfter timeoutAfter:Int,
         withTimeoutHandler handler:(() -> Void)?) {
-            guard status != SocketIOClientStatus.Connected else {
+            assert(timeoutAfter >= 0, "Invalid timeout: \(timeoutAfter)")
```
Don't use assert

NachoSoto replied:
Why not? It doesn't make any sense to pass a negative `timeout`. What should be the expected behavior? The `dispatch_after` would be invoked right away so it would timeout right away.

lightsprint09 replied:
would it make sense to change the parameter type from `Int` to `UInt`

NachoSoto replied:
Apple is against unsigned types (they've almost removed them from all their APIs), but I do agree with that solution, specially since Swift guards against overflows.

nuclearace replied:
I'm also against unsigned types. Just use checks or treat negatives as 0

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -148,28 +153,30 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
             
             let time = dispatch_time(DISPATCH_TIME_NOW, Int64(timeoutAfter) * Int64(NSEC_PER_SEC))
-            
-            dispatch_after(time, dispatch_get_main_queue()) {[weak self] in
```
Again, I don't think it's good to keep strong references like this

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -181,15 +188,15 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
             if let this = self {
                 this.ackHandlers.addAck(ack, callback: callback)
                 
-                dispatch_async(this.emitQueue) {[weak this] in
```
There's no reason this should change to a strong reference

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -181,15 +188,15 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
                     
-                    dispatch_after(time, dispatch_get_main_queue()) {[weak this] in
-                        this?.ackHandlers.timeoutAck(ack)
+                    dispatch_after(time, dispatch_get_main_queue()) {
```
Should be a weak `this`

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -249,12 +256,12 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
-        dispatch_async(emitQueue) {[weak self] in
-            self?._emit([event] + items)
+
+        dispatch_async(emitQueue) {
```
`weak self` is fine here

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -292,17 +299,17 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
     
     // If the server wants to know that the client received data
     func emitAck(ack: Int, withItems items: [AnyObject]) {
-        dispatch_async(emitQueue) {[weak self] in
```
There's not really a reason for this to be strong.

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -334,22 +341,21 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
             if anyHandler != nil {
-                dispatch_async(handleQueue) {[weak self] in
-                    self?.anyHandler?(SocketAnyEvent(event: event, items: data))
+                dispatch_async(handleQueue) {
```
Should be `weak self`

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -334,22 +341,21 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
             for handler in handlers where handler.event == event {
                 if let ack = ack {
-                    dispatch_async(handleQueue) {[weak self] in
+                    dispatch_async(handleQueue) {
```
Should be `weak self`

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -441,18 +447,14 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
-            if let this = self {
-                SocketParser.parseSocketMessage(msg, socket: this)
-            }
+        dispatch_async(handleQueue) {
```
Should be `weak self`

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -441,18 +447,14 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
-            if let this = self {
-                SocketParser.parseBinaryData(data, socket: this)
-            }
+        dispatch_async(handleQueue) {
```
Should be `weak self`

nuclearace commented on SocketIOClientSwift/SocketIOClient.swift:
```diff
@@ -468,19 +470,17 @@ public final class SocketIOClient: NSObject, SocketEngineClient {
-                    this.reconnectTimer = NSTimer.scheduledTimerWithTimeInterval(Double(this.reconnectWait),
-                        target: this, selector: "_tryReconnect", userInfo: nil, repeats: true)
-                }
+            dispatch_async(dispatch_get_main_queue()) {
```
Should be `weak self`

nuclearace commented on SocketIOClientSwift/SocketPacket.swift:
```diff
@@ -28,6 +28,8 @@ struct SocketPacket {
     private let placeholders: Int
     private var currentPlace = 0
 
+	private static let logType = "SocketPacket"
```
Indentation is a little wonky in this file

NachoSoto replied:
Sorry! I'll fix

nuclearace commented on SocketIOClientSwift/WebSocket.swift:
```diff
@@ -378,21 +378,17 @@ public class WebSocket : NSObject, NSStreamDelegate {
         }
         if totalSize > 0 {
             if validateResponse(buffer, bufferLen: totalSize) {
-                dispatch_async(queue, { [weak self] in
```
I believe these weak references are on purpose, otherwise there's a case where the WebSocket won't be released.

NachoSoto replied:
This strong reference is only kept for a short period of time (while this block is invoked). This **won't** cause a retain cycle. I'll explain more separately though.

nuclearace commented:
Two things. Generally I take the approach that unless there's an explicit reason to perform a strong capture, always perform a weak capture. This way you avoid crazy cases where you've got a strong reference that's keeping your object alive.

Also, the weak references in the WebSocket are intentional. https://github.com/daltoniam/Starscream/commit/889d2880eb4585373b1bc3907574399347d440ff

hankbao commented:
@nuclearace There is a race condition in `WebSocket.connect()`. Please see my [comment](https://github.com/socketio/socket.io-client-swift/pull/170#issuecomment-139485493) above.

I would like to open an issue about that if you think that is appropriate.

nuclearace commented:
@hankbao I thought you fixed that in the thing I merged yesterday? Is it still present in the default [WebSocket](https://github.com/daltoniam/Starscream)? I generally try to back-patch those sort of things. If it's still an issue there I'd make an issue there.

hankbao commented:
@nuclearace IMO, the multithreads race condition in `connect` method is different from the use-after-release problem caused by a unowned object and that is not fixed by weak-referencing self. After turned on forceWebsocket, it's now easy to reproduce the problem on iOS 9 GM if you connect-close-connect quickly.

nuclearace commented:
You're welcome to provide a fix, I'm having trouble reproducing it on the swift-2 branch. But I do most of my testing on OS X.

lightsprint09 commented:
@hankbao it would be great to create a test case to test against automatically

hankbao commented:
@nuclearace @lightsprint09 OK. But I'm quite busy currently and it's almost at midnight in China. I'd rather work on this issue tomorrow.

lightsprint09 commented:
I think we are all in different time zones :+1:

nuclearace commented:
@hankbao That's fine. I don't think this is really a major issue

NachoSoto commented:
> @nuclearace IMO, the multithreads race condition in connect method is different from the use-after-release problem caused by a unowned object and that is not fixed by weak-referencing self. After turned on forceWebsocket, it's now easy to reproduce the problem on iOS 9 GM if you connect-close-connect quick

That's right, there's a very big difference between unowned and weak. In that sense, something like [this commit](https://github.com/daltoniam/Starscream/commit/889d2880eb4585373b1bc3907574399347d440ff) is right because `unowned` can crash the app.

> let weakSelf = self will try to unwrap and retain a deallocated object and eventually crashes the app.

I'm sorry, but that's completely wrong. if `self` is a weak reference, doing `if let weakSelf = self` won't crash the app. `self` will have been nil-ed out (because it's `weak`, not `unowned`), which means that doing `if-let` simply defines `weakSelf` **only if** it's not `nil`. That's completely say.

>Two things. Generally I take the approach that unless there's an explicit reason to perform a strong capture, always perform a weak capture. This way you avoid crazy cases where you've got a strong reference that's keeping your object alive.

Using `weak` is useful for places where you want to avoid a retain cycle, like [here](https://github.com/socketio/socket.io-client-swift/blob/swift-2/SocketIOClientSwift/SocketIOClient.swift#L180).

But all the ones I've removed in this PR are unnecessary:
- `dispatch_*` is guaranteed to be invoked some time in the future, which means that the cycle will always be broken.
- `weak` introduces an unnecessary performance hit: there needs to be extra work to monitor references to them `nil-out`.
- `weak` introduces unnecessary complexity: you need to consider in each code path what the right behavior should be if `self` is dealloced. We can **simplify** the code by avoiding this.
- `weak` introduces more code than needed: look at how much code I removed: `weak self`, `guard`, etc. Code becomes a lot simpler, while keeping the same behavior.

NachoSoto commented:
Thanks for the comments guys!

nuclearace commented:
>dispatch_* is guaranteed to be invoked some time in the future, which means that the cycle will always be broken.

I wasn't aware of that

nuclearace commented:
I would prefer to keep the WebSockets library as is. If you think the `weak`s are uneeded, make an issue on that [repo](https://github.com/daltoniam/Starscream) I generally try to keep up that

NachoSoto commented:
Why would an issue be more helpful? I've already done the work for you.

Is there a particular reason you disagree with this other than wanting to keep `weak` because you don't understand it?

What about the rest of changes?

nuclearace commented:
Taking the `weak`s out is fine if what you say about dispatch_* always executing the closure is true. I just wasn't sure about the WebSocket one, since I didn't write it.

NachoSoto commented:
Cool. Which one aren't you sure about? Maybe one of them is actually needed.

NachoSoto commented:
It's not *just* for the optimization, it's mostly simplifying the code to make it easier to maintain :)

NachoSoto commented:
And because it means we no longer have to think about what should happen if self is nil everywhere.

nuclearace commented:
The ones in https://github.com/daltoniam/Starscream/commit/889d2880eb4585373b1bc3907574399347d440ff Not sure if you removed some of these.

nuclearace commented:
And fyi, I'm all for taking the `weak`s out, I just wasn't sure if there's going to be some edge case where there's a strong retain that's keeping the client from being released.

nuclearace commented on SocketIOClientSwift/SocketEngine.swift:
```diff
@@ -495,7 +497,7 @@ public final class SocketEngine: NSObject, WebSocketDelegate {
         var msg = ""
 
-        func testLength(length:String, inout n:Int) -> Bool {
+        func testLength(length: String, inout n: Int) -> Bool {
```
Also thanks, I keep finding places where I've forgotten to remove the old style.

NachoSoto commented:
> The ones in daltoniam/Starscream@889d288 Not sure if you removed some of these.

I haven't :)

> And fyi, I'm all for taking the weaks out, I just wasn't sure if there's going to be some edge case where there's a strong retain that's keeping the client from being released.

I studied each of them very rigorously (I didn't just got rid of all `weak`s). I only removed the ones where it's guaranteed that the cycle will be broken. In the case of `dispatch`, the block disappears (and with it, its `strong` `self` reference) as soon as it's invoked. It's impossible that that will create a cycle. There's actually never a cycle because `self` doesn't own the block.

NachoSoto commented:
Thanks :star:
